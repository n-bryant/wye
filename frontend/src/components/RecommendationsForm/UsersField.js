import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Icon from "@mdi/react";
import { mdiAccountPlus, mdiAccountPlusOutline } from "@mdi/js";
import Chip from "@material-ui/core/Chip";

import styles from "./UsersField.styles";
import Typography from "@material-ui/core/Typography";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";

// setting a max user count for performance gains
export const MAX_USER_COUNT = 6;

/**
 * renders a TextField that allows for inputting
 * multiple SteamID values that are rendered as removable Chips
 */
export const UsersField = props => {
  const classnames = UsersField.classnames(props);
  const { classes, name, setFieldValue, value } = props;
  const [textFieldValue, setTextFieldValue] = React.useState("");

  /**
   * adds a user to the users field value
   * @param {String} user
   */
  const addUser = user => {
    setFieldValue(name, value.concat([user]));
    setTextFieldValue("");
  };

  /**
   * removes a user from the users field value at the specified index
   * @param {Int} index
   */
  const removeUser = index => {
    setFieldValue(
      name,
      value.filter((_el, elIndex) => elIndex !== index)
    );
  };

  // should not be able to add a user if the text field is blank,
  // or if the allowed number of users has already been added
  const addUserDisabled =
    !textFieldValue.length ||
    value.length >= MAX_USER_COUNT ||
    value.some(val => val === textFieldValue);
  const textFieldDisabled = value.length >= MAX_USER_COUNT;

  return (
    <div className={classnames.root()}>
      <div className={classnames.element("addUserfieldContainer")}>
        <TextField
          className={classnames.element("addUserField")}
          InputLabelProps={{
            classes: {
              disabled: props.classes.fieldRoot
            }
          }}
          label="Add a SteamID:"
          placeholder="e.g. - 76561197975995523"
          autoFocus={true}
          disabled={textFieldDisabled}
          value={textFieldValue}
          onChange={event => {
            setTextFieldValue(event.target.value);
          }}
          onKeyPress={event => {
            // allow for usage of the enter key to add a user
            if (event.key === "Enter") {
              event.preventDefault();

              if (!addUserDisabled) {
                addUser(event.target.value);
              }
            }
          }}
        />
        <Button
          className={classnames.element("addUserButton")}
          onClick={() => {
            addUser(textFieldValue);
          }}
          disabled={addUserDisabled}
        >
          <Icon
            className={classnames.element("addUserIcon", {
              enabled: !addUserDisabled,
              disabled: addUserDisabled
            })}
            path={
              !addUserDisabled && textFieldValue.length
                ? mdiAccountPlus
                : mdiAccountPlusOutline
            }
          />
        </Button>
      </div>
      <div className={classnames.element("fieldItemsContainer")}>
        {value &&
          value.map((_user, index) => {
            return (
              <Chip
                key={index}
                label={value[index]}
                onDelete={() => removeUser(index)}
                className={classnames.element("chip")}
                variant="outlined"
                classes={{
                  outlined: classes.chip,
                  deleteIcon: classes.deleteIcon
                }}
              />
            );
          })}
      </div>
    </div>
  );
};
UsersField.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}UsersField`
);
UsersField.propTypes = {
  // styles to apply to the form
  classes: PropTypes.shape({
    root: PropTypes.string,
    fieldItemsContainer: PropTypes.string,
    chip: PropTypes.string,
    deleteIcon: PropTypes.string,
    addUserfieldContainer: PropTypes.string,
    addUserField: PropTypes.string,
    addUserButton: PropTypes.string,
    addUserIcon: PropTypes.string,
    addUserIconEnabled: PropTypes.string
  }),
  // Formik field name
  name: PropTypes.string,
  // users field value from Formik form
  value: PropTypes.array,
  // handler for updating users field value
  setFieldValue: PropTypes.func
};
UsersField.defaultProps = {
  classes: {}
};

// apply styles
export default withStyles(styles)(UsersField);
