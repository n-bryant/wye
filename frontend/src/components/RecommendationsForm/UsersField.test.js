import { mdiAccountPlus } from "@mdi/js";
import Chip from "@material-ui/core/Chip";
import { UsersField } from "./UsersField";

describe("UsersField", () => {
  const props = {
    classes: {
      root: "root",
      fieldItemsContainer: "fieldItemsContainer",
      chipsLabel: "chipsLabel",
      chipsLabelHidden: "chipsLabelHidden",
      chip: "chip",
      deleteIcon: "deleteIcon",
      addUserfieldContainer: "addUserfieldContainer",
      addUserField: "addUserField",
      addUserButton: "addUserButton",
      addUserIcon: "addUserIcon",
      addUserIconEnabled: "addUserIconEnabled",
      addUserIconDisabled: "addUserIconDisabled"
    },
    name: "field",
    value: [],
    setFieldValue: jest.fn()
  };

  const setTextFieldValue = jest.fn();

  let wrapper;
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<UsersField {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set the TextField disabled value to true if the length of the usersValue prop is greater than or equal to the max user count", () => {
    const textField = wrapper.findWhere(n =>
      n.hasClass(props.classes.addUserField)
    );
    expect(textField.prop("disabled")).toBeFalsy();

    wrapper.setProps({
      ...props,
      value: ["a", "b", "c", "d", "e", "f"]
    });
    wrapper.update();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.addUserField))
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should set the TextField value prop to the textFieldValue state value", () => {
    const textFieldValue = "foo";
    const setTextFieldValue = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);
    const textField = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserField)
    );
    expect(textField.prop("value")).toBe(textFieldValue);
  });

  it("should call the setTextFieldValue state update handler with the event target's value when the TextField experiences a change", () => {
    const textFieldValue = "foo";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);
    const textField = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserField)
    );
    const newValue = "bar";
    textField.prop("onChange")({ target: { value: newValue } });
    expect(setTextFieldValue).toHaveBeenCalledWith(newValue);
  });

  it("should add a user and reset the TextField when the Enter key is pressed and Textfield is not disabled", () => {
    const textFieldValue = "foo";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);
    const textField = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserField)
    );

    const newValue = "bar";
    textField.prop("onKeyPress")({
      key: "Enter",
      target: { value: newValue },
      preventDefault: jest.fn()
    });
    expect(props.setFieldValue).toHaveBeenCalledWith(props.name, [newValue]);
  });

  it("should add a user when the add user button is clicked", () => {
    const textFieldValue = "foo";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);
    const addUserButton = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserButton)
    );
    addUserButton.prop("onClick")();
    expect(props.setFieldValue).toHaveBeenCalledWith(props.name, [
      textFieldValue
    ]);
  });

  it("should set the add user button to disabled if the text field value is blank", () => {
    const textFieldValue = "";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);
    const addUserButton = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserButton)
    );
    expect(addUserButton.prop("disabled")).toBeTruthy();
  });

  it("should set the add user button to disabled if the max user count has been reached", () => {
    const newProps = {
      ...props,
      value: ["a", "b", "c", "d", "e", "f"]
    };
    const textFieldValue = "foo";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...newProps} />);
    const addUserButton = wrapperWithState.findWhere(n =>
      n.hasClass(newProps.classes.addUserButton)
    );
    expect(addUserButton.prop("disabled")).toBeTruthy();
  });

  it("should set enabled styles to the add user icon if the text field is not blank", () => {
    const textFieldValue = "foo";
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementationOnce(() => [
      textFieldValue,
      setTextFieldValue
    ]);

    const wrapperWithState = shallow(<UsersField {...props} />);

    const addUserIcon = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.addUserIconEnabled)
    );
    expect(addUserIcon.prop("path")).toBe(mdiAccountPlus);
  });

  it("should render a Chip for each added user", () => {
    const value = ["a", "b", "c"];
    wrapper.setProps({
      value
    });
    wrapper.update();

    expect(wrapper.find(Chip).length).toBe(value.length);
  });

  it("should set the Chips' label to the added user's SteamID", () => {
    const value = ["a", "b", "c"];
    wrapper.setProps({
      value
    });
    wrapper.update();
    wrapper.find(Chip).forEach((chip, index) => {
      expect(chip.prop("label")).toBe(value[index]);
    });
  });

  it("should set outlined and deleteIcon classes for the Chips", () => {
    const value = ["a"];
    wrapper.setProps({
      value
    });
    wrapper.update();
    expect(wrapper.find(Chip).prop("classes")).toEqual({
      outlined: props.classes.chip,
      deleteIcon: props.classes.deleteIcon
    });
  });
});
