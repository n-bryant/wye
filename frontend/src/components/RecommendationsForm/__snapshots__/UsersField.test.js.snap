// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersField should render successfully 1`] = `
<div
  className="wye-UsersField root"
>
  <div
    className="wye-UsersField__addUserfieldContainer addUserfieldContainer"
  >
    <WithStyles(ForwardRef(TextField))
      InputLabelProps={
        Object {
          "classes": Object {
            "disabled": undefined,
          },
        }
      }
      autoFocus={true}
      className="wye-UsersField__addUserField addUserField"
      disabled={false}
      label="Add a SteamID:"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="e.g. - 76561197975995523"
      value=""
    />
    <WithStyles(ForwardRef(Button))
      className="wye-UsersField__addUserButton addUserButton"
      disabled={true}
      onClick={[Function]}
    >
      <Icon
        className="wye-UsersField__addUserIcon wye-UsersField__addUserIcon--disabled addUserIcon addUserIconDisabled"
        color="currentColor"
        horizontal={false}
        path="M15,4A4,4 0 0,0 11,8A4,4 0 0,0 15,12A4,4 0 0,0 19,8A4,4 0 0,0 15,4M15,5.9C16.16,5.9 17.1,6.84 17.1,8C17.1,9.16 16.16,10.1 15,10.1A2.1,2.1 0 0,1 12.9,8A2.1,2.1 0 0,1 15,5.9M4,7V10H1V12H4V15H6V12H9V10H6V7H4M15,13C12.33,13 7,14.33 7,17V20H23V17C23,14.33 17.67,13 15,13M15,14.9C17.97,14.9 21.1,16.36 21.1,17V18.1H8.9V17C8.9,16.36 12,14.9 15,14.9Z"
        rotate={0}
        size={null}
        spin={false}
        vertical={false}
      />
    </WithStyles(ForwardRef(Button))>
  </div>
  <div
    className="wye-UsersField__fieldItemsContainer fieldItemsContainer"
  />
</div>
`;
